### PR TITLE
Set XDEBUG_MODE for processes with coverage 

### LINKED ISSUE #1473: For recent Xdebug coverage collection needs XDEBUG_MODE=coverage or xdebug.mode=coverage
| Question    | Answer
| ------------| ---------------
| Infection version | 0.20.2
| Test Framework version | PHPUnit 9.5.0
| PHP version | PHP 7.4.14 (cli) with Xdebug v3.0.2
| Platform    | Docker (php:7.4-fpm)
| Github Repo | https://github.com/jenshaase/phel-lang

</details>

<details>
 <summary>Output with issue</summary>

 ```

 dev@php:/srv/phel-lang$ vendor/bin/infection

    ____      ____          __  _
   /  _/___  / __/__  _____/ /_(_)___  ____
   / // __ \/ /_/ _ \/ ___/ __/ / __ \/ __ \
 _/ // / / / __/  __/ /__/ /_/ / /_/ / / / /
/___/_/ /_/_/  \___/\___/\__/_/\____/_/ /_/

Infection - PHP Mutation Testing Framework version 0.20.2@6035c1566af6a5a8d833a276351e5e18ed412305


Running initial test suite...

PHPUnit version: 9.5.0

  614 [============================]  1 min

 In CoverageChecker.php line 169:

  The code coverage generated by the initial test run is invalid. Please report the issue on the
  infection repository "https://github.com/infection/infection".

  $ '/srv/phel-lang/vendor/phpunit/phpunit/phpunit' '--configuration' '/srv/phel-lang/data/infection/infection/
  phpunitConfiguration.initial.infection.xml' '-vvv' '--coverage-xml=/srv/phel-lang/data/infection/infection/co
  verage-xml' '--log-junit=/srv/phel-lang/data/infection/infection/junit.xml'
  PHPUnit 9.5.0 by Sebastian Bergmann and contributors.

  Runtime:       PHP 7.4.14
  Configuration: /srv/phel-lang/data/infection/infection/phpunitConfiguration.initial.infection.xml
  Random Seed:   1611098728
  Warning:       XDEBUG_MODE=coverage or xdebug.mode=coverage has to be set

  ...............................................................  63 / 608 ( 10%)
  ............................................................... 126 / 608 ( 20%)
  ............................................................... 189 / 608 ( 31%)
  ............................................................... 252 / 608 ( 41%)
  ............................................................... 315 / 608 ( 51%)
  ............................................................... 378 / 608 ( 62%)
  ............................................................... 441 / 608 ( 72%)
  ............................................................... 504 / 608 ( 82%)
  ............................................................... 567 / 608 ( 93%)
  .........................................                       608 / 608 (100%)

  Time: 01:03.278, Memory: 28.00 MB

  OK (608 tests, 861 assertions)


  Issue(s):
  - The file "index.xml" could not be found: Could not find any "index.xml" file in "/srv/phel-lang/data/infection/infection"
 ```
</details>

---

I got the following error:
```
The file "index.xml" could not be found: Could not find any "index.xml" file in 
"/srv/phel-lang/data/infection/infection"
```

I have been trying to find this issue but the only one seems related is this one: https://github.com/infection/infection/issues/30
But it's too old and I think it's not exactly the same issue.

You can take a look at the project if you want, there you are everything.

Do you have any clue about this? 😄

### PR DESCRIPTION
**Pull request by sanmai**
This PR:

- [x] Follow up to #1510
- [x] Set `XDEBUG_MODE` for processes with coverage so a user doesn't even have to know this knob exists
- [ ] Covered by tests
- [x] Fixes #1473

Testing steps:

- Note how build failed on aa1e089 when we disabled Xdebug

Implementation notes:

 - PHPUnit 9.4.2 doesn't really work well with Xdebug 3 (even with `XDEBUG_MODE=coverage` it emits a warning "xdebug.mode=coverage has to be set in php.ini" and fails to generate coverage). Upgrading to 9.5.10 fixes the problem.

### CODE REVIEW

File: src/Process/CoveredPhpProcess.php [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+ * required when PCOV already used for coverage.
+ */
+final class CoveredPhpProcess extends Process
+{
+    /**
+     * @param array<string|bool>|null $env
+     */
+    public function start(?callable $callback = null, ?array $env = null): void
+    {
+        if (!extension_loaded('pcov')) {
```
Comment (sidz) [inline src/Process/CoveredPhpProcess.php:60]: we also support phpdbg

File: src/Process/CoveredPhpProcess.php [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            $env = array_merge($env ?? [], [
+                'XDEBUG_MODE' => 'debug',
```
Comment (maks-rafalko) [inline src/Process/CoveredPhpProcess.php:62]: could you please explain why `debug` but not `coverage`?

https://xdebug.org/docs/all_settings#mode 

From what I understand, and from what I see in my tests, `debug` provide coverage report, while `coverage` does.

* `XDEBUG_MODE=debug bin/infection ...` - fails
* * `XDEBUG_MODE=coverage bin/infection ...` - works

Am I missing something?

File: src/Process/CoveredPhpProcess.php (hunk 1 above)
Comment (sanmai) [reply src/Process/CoveredPhpProcess.php:60]: I can add a check for it, but they are mutually exclusive. You can't use Xdebug with PHPDBG, IIRC, so whenever we export or not this variable does not matter. 

On the contrary, you can have both Xdebug and PCOV. And that's why we're making sure of not breaking stuff.

File: src/Process/CoveredPhpProcess.php (hunk 2 above)
Comment (sanmai) [reply src/Process/CoveredPhpProcess.php]: No, you're 100% right. This should be `XDEBUG_MODE=coverage`.

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -24,6 +24,7 @@ jobs:
           - { operating-system: 'windows-latest', php-version: '7.4', dependencies: '', coverage-driver: 'xdebug' }
           - { operating-system: 'ubuntu-latest', php-version: '7.4', dependencies: '--prefer-lowest', coverage-driver: 'pcov' }
           - { operating-system: 'ubuntu-latest', php-version: '8.0', dependencies: '--ignore-platform-req=php', coverage-driver: 'pcov' }
+          - { operating-system: 'ubuntu-latest', php-version: '7.4', dependencies: '', coverage-driver: 'xdebug2' }
```
Comment (sanmai) [inline .github/workflows/ci.yaml:27]: Need to verify we're not already using Xdebug 2.

**Comment by sanmai** [comment]
I think I messed up something here. Let me take a day to check everything again.

File: src/Process/CoveredPhpProcess.php (hunk 1 above)
Comment (sidz) [reply src/Process/CoveredPhpProcess.php:60]: I'm voting to add add phpdbg check to this condition.

File: tests/phpunit/Process/OriginalPhpProcessTest.php [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    public function test_it_injects_xdebug_env_vars(): void
+    {
+        $process = new OriginalPhpProcess(['env']);
+        $process->run(null, ['TESTING' => 'test']);
+
+        if (
+            extension_loaded('pcov') ||
+            PHP_SAPI === 'phpdbg' ||
+            XdebugHandler::getSkippedVersion() !== '' ||
+            ini_get_unsafe('xdebug.mode') !== false
```
[the remark this answers is not in the document]
Comment (sanmai) [reply tests/phpunit/Process/OriginalPhpProcessTest.php:73]: Three weeks later DI stuff isn't here, but it looks like I was able to figure it out: here we assume that 
`XDEBUG_MODE=coverage` is set unless we're running with PCOV, or under PHPDBG, or with Xdebug <3. Checking for `XdebugHandler` here is a foul business: we never run invoke it for tests.

File: src/Process/OriginalPhpProcess.php [hunk 5]
```diff
@@ -57,6 +62,22 @@ public function start(?callable $callback = null, ?array $env = null): void
...
+            // it too, but it has coverage enabled at all times.
+            ini_get_unsafe('xdebug.mode') !== false
+        ) {
+            // Why going through all the trouble above? We don't want to enable
+            // Xdebug when there are more compelling choices. In the end the user is
+            // still in control: they can provide XDEBUG_MODE=coverage on their own.
+            $env = array_merge($env ?? [], [
+                'XDEBUG_MODE' => 'coverage',
+            ]);
+        }
```
[the remark this answers is not in the document]
Comment (sanmai) [reply src/Process/OriginalPhpProcess.php:69]: OK, all should be much more clear in `shallExtendEnvironmentWithXdebugMode` below.

**Review by sanmai — COMMENTED** [review]
:crossed_fingers:

**Review by maks-rafalko — APPROVED** [review]
:clap: looks more clear now

**Comment by alfredbez** [comment]
I'm getting an error since this change was merged and released:
```
Infection - PHP Mutation Testing Framework version 0.25.4.0


Fatal error: Uncaught Error: Call to undefined method Composer\XdebugHandler\XdebugHandler::isXdebugActive() in phar:///usr/local/bin/infection/src/TestFramework/Coverage/CoverageChecker.php:99
Stack trace:
#0 phar:///usr/local/bin/infection/src/TestFramework/Coverage/CoverageChecker.php(47): _HumbugBox75a5b98e45f1\Infection\TestFramework\Coverage\CoverageChecker->hasCoverageGeneratorEnabled()
#1 phar:///usr/local/bin/infection/src/Command/RunCommand.php(140): _HumbugBox75a5b98e45f1\Infection\TestFramework\Coverage\CoverageChecker->checkCoverageRequirements()
#2 phar:///usr/local/bin/infection/src/Command/RunCommand.php(72): _HumbugBox75a5b98e45f1\Infection\Command\RunCommand->startUp(Object(_HumbugBox75a5b98e45f1\Infection\Container), Object(_HumbugBox75a5b98e45f1\Infection\Console\ConsoleOutput), Object(_HumbugBox75a5b98e45f1\Infection\Logger\ConsoleLogger), Object(_HumbugBox75a5b98e45f1\Infection\Console\IO))
#3 phar:///usr/local/bin/infection/src/Command/BaseCommand.php(28): _HumbugBox75a5b98e45f1\Infection\Command\Ru in phar:///usr/local/bin/infection/src/TestFramework/Coverage/CoverageChecker.php on line 99
```

My `Dockerfile` looks like this:
```
FROM mileschou/xdebug:7.4-alpine

RUN docker-php-ext-install bcmath

RUN apk add --no-cache git

RUN wget https://github.com/infection/infection/releases/download/0.25.4/infection.phar \
    && wget https://github.com/infection/infection/releases/download/0.25.4/infection.phar.asc \
    && chmod +x infection.phar \
    && mv infection.phar /usr/local/bin/infection

COPY --from=composer /usr/bin/composer /usr/bin/composer
```

The same behaviour occurs in all future releases after 0.25.4.
Am I doing something wrong?

**Pull request merged**